### PR TITLE
[FlexibleHeader] Fixes to support scroll views with Safe Area insets.

### LINKED ISSUE #2008: [FlexibleHeader] Support Safe Area
### Overview
Issues visible in the Catalog: Multiple examples have issues with the default header height (taller than expected). The Standard UINavigationBar example has issues with the positioning of the buttons and the status bar. The Configurator example has issues when Shift Behavior is enabled, as described in #1969.

#### Configurator:
configurator

configurator

#### Standard UINavigationBar
buttons

### Version number
35.0.0

### Operating system and device
iOS 11 / iPhone X

### PR DESCRIPTION
**Pull request by chuga**
### TLDR
This PR adds support for Safe Area insets in Flexible Header by adjusting injectedTopContentInset to account for any Safe Area insets the tracking scroll view might already come with. It also removes all the hardcoded status bar height values. This PR doesn't introduce new API, things should just work as expected.

### Why this change
If the scroll view we're tracking has contentInsetAdjustmentBehavior set to Automatic (the default) or Always, it already comes with an additional inset to account for the Safe Area. Since we don't own these scroll views, and these modes are actually pretty useful as they take care of the Safe Area for you, we can't just turn it off when they're added to the flexible header. The right way to fix it is to take this inset and adjust how much we need to add on top of it to match the height of the header view.

### Default height
The original default height for the flexible header was hardcoded to 56 + 20 (for the status bar).
Since the status bar now can be 44, 20 or 0, the default height should be adjusted accordingly. If a developer is currently setting the height manually, we don't do any adjustments on our end. We should consider adding a flag that might enable that functionality (adding the status bar dynamically), but that should be on a separate PR.

### Testing
This fix has been manually tested on an iPhone X and 6 (both simulators), on iOS 9.3, 10.3 and 11.0 using all the examples we currently maintain in the Catalog for this component.

Closes #2008 and #2004.

### CODE REVIEW

File: components/AppBar/src/MDCAppBar.m [hunk 1]
```diff
@@ -33,7 +34,6 @@
 static NSString *const MDCAppBarHeaderViewControllerKey = @"MDCAppBarHeaderViewControllerKey";
 static NSString *const MDCAppBarNavigationBarKey = @"MDCAppBarNavigationBarKey";
 static NSString *const MDCAppBarHeaderStackViewKey = @"MDCAppBarHeaderStackViewKey";
-static const CGFloat kStatusBarHeight = 20;
```
Comment (jverkoey) [inline components/AppBar/src/MDCAppBar.m:36]: 🎉

File: components/FlexibleHeader/src/MDCFlexibleHeaderView.m [hunk 2]
```diff
@@ -16,16 +16,15 @@
 
 #import "MDCFlexibleHeaderView.h"
 
+#import "MaterialApplication.h"
```
Comment (jverkoey) [inline components/FlexibleHeader/src/MDCFlexibleHeaderView.m:19]: This looks out of place - remove?

File: components/FlexibleHeader/src/MDCFlexibleHeaderView.m [hunk 3]
```diff
@@ -109,6 +108,12 @@ @interface MDCFlexibleHeaderScrollViewInfo : NSObject
 // The amount injected into scrollIndicatorInsets.top
 @property(nonatomic) CGFloat injectedTopScrollIndicatorInset;
 
+// The adjustment we've made to account for the scroll view's Safe Area.
+@property(nonatomic) CGFloat topSafeAreaInsetAdjustment;
+
+// Whether or not we've adjust the inset to account for the scroll view's Safe Area.
```
Comment (jverkoey) [inline components/FlexibleHeader/src/MDCFlexibleHeaderView.m]: typo: adjusted

Comment (ianegordon) [inline components/FlexibleHeader/src/MDCFlexibleHeaderView.m]: s/adjust/adjusted/

File: components/FlexibleHeader/src/MDCFlexibleHeaderViewController.m [hunk 4]
```diff
@@ -16,6 +16,7 @@
 
 #import "MDCFlexibleHeaderViewController.h"
 
+#import "MaterialApplication.h"
```
Comment (jverkoey) [inline components/FlexibleHeader/src/MDCFlexibleHeaderViewController.m:19]: Remove?

File: components/FlexibleHeader/src/MDCFlexibleHeaderView.m (hunk 2 above)
Comment (jverkoey) [reply components/FlexibleHeader/src/MDCFlexibleHeaderView.m:19]: Alternatively, if it's required, add `    ss.dependency "MaterialComponents/private/Application"` to the podspec.

**Review by jverkoey — CHANGES_REQUESTED** [review]

File: components/FlexibleHeader/src/MDCFlexibleHeaderView.m (hunk 2 above)
Comment (chuga) [reply components/FlexibleHeader/src/MDCFlexibleHeaderView.m:19]: We need it in order to get the status bar height ([UIApplication mdc_safeSharedApplication])

Comment (chuga) [reply components/FlexibleHeader/src/MDCFlexibleHeaderView.m:19]: Looks like it's already there? https://github.com/material-components/material-components-ios/blob/develop/MaterialComponents.podspec#L255

File: components/FlexibleHeader/src/MDCFlexibleHeaderViewController.m (hunk 4 above)
Comment (chuga) [reply components/FlexibleHeader/src/MDCFlexibleHeaderViewController.m:19]: Ditto

**Comment by chuga** [comment]
Bug: Since the height of the header depends on the status bar height (unless it was explicitly set to something else), I need to change the size of the header when the safe area / status bar height changes (like when you rotate the device). I'll be looking into that tomorrow, but feel free to review this in the meantime.

**Review by ajsecord — APPROVED** [review]

**Comment by romoore** [comment]
Was this tested on Xcode 8/iOS 8 simulators? From the summary, it looks like it was tested using Xcode 9GM?

File: catalog/MDCCatalog/Info.plist [hunk 5]
```diff
@@ -3,7 +3,7 @@
 <plist version="1.0">
 <dict>
 	<key>UIViewControllerBasedStatusBarAppearance</key>
-	<false/>
+	<true/>
```
Comment (romoore) [inline catalog/MDCCatalog/Info.plist]: Nit: Remove?
 
> When this key is not present or its value is set to YES, the view controller determines the status bar style.
&mdash; [Apple Docs](https://developer.apple.com/library/content/documentation/General/Reference/InfoPlistKeyReference/Articles/iPhoneOSKeys.html#//apple_ref/doc/uid/TP40009252-SW10)

**Comment by chuga** [comment]
Not yet, but we should before we merge it in.

File: components/FlexibleHeader/examples/FlexibleHeaderHorizontalPagingExample.m [hunk 6]
```diff
@@ -146,6 +146,7 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder {
 
 - (void)commonMDCFlexibleHeaderViewControllerInit {
   self.fhvc = [[MDCFlexibleHeaderViewController alloc] initWithNibName:nil bundle:nil];
+  self.fhvc.headerView.sharedWithManyScrollViews = YES;
```
Comment (romoore) [inline components/FlexibleHeader/examples/FlexibleHeaderHorizontalPagingExample.m:149]: Curiosity: Why does this need to change?  Was the behavior previously broken for this example?

File: components/FlexibleHeader/examples/supplemental/FlexibleHeaderTopLayoutGuideSupplemental.m [hunk 7]
```diff
@@ -49,7 +49,7 @@ - (void)setupScrollViewContent {
   UIColor *color =
       [UIColor colorWithRed:(97.0 / 255.0) green:(97.0 / 255.0) blue:(97.0 / 255.0) alpha:1.0];
   UIView *scrollViewContent =
-      [[UIView alloc] initWithFrame:CGRectMake(0, 0, self.scrollView.frame.size.width, 700)];
+      [[UIView alloc] initWithFrame:CGRectMake(0, 0, self.scrollView.frame.size.width, 1000)];
```
Comment (romoore) [inline components/FlexibleHeader/examples/supplemental/FlexibleHeaderTopLayoutGuideSupplemental.m]: Nit/suggestion: Does it make more sense to grab the main screen and use double the max of its height/width or something?  It seems like this is a fix due to iPhone X being taller than iPhone *Plus.  Does this still scroll on an iPad Pro (1366 points)?

File: components/FlexibleHeader/src/MDCFlexibleHeaderView.m [hunk 8]
```diff
@@ -109,6 +108,12 @@ @interface MDCFlexibleHeaderScrollViewInfo : NSObject
 // The amount injected into scrollIndicatorInsets.top
 @property(nonatomic) CGFloat injectedTopScrollIndicatorInset;
 
+// The adjustment we've made to account for the scroll view's Safe Area.
+@property(nonatomic) CGFloat topSafeAreaInsetAdjustment;
+
+// Whether or not we've adjusted the inset to account for the scroll view's Safe Area.
+@property(nonatomic) BOOL hasTopSafeAreaInsetAdjustment;
```
Comment (romoore) [inline components/FlexibleHeader/src/MDCFlexibleHeaderView.m:117]: Nit (BOOL property getter): 
```objectivec
@property(nonatomic, getter=isTopSafeAreaInsetAdjusted) BOOL topSafeAreaInsetAdjusted;
```

File: components/FlexibleHeader/src/private/MDCStatusBarShifter.m [hunk 9]
```diff
@@ -56,21 +55,38 @@ @implementation MDCStatusBarShifter {
...
+               name:UIApplicationDidChangeStatusBarFrameNotification
+             object:nil];
   }
   return self;
 }
 
+#pragma mark - Notification
+
+- (void)statusBarDidChangeFrame {
+  CGFloat statusBarHeight = [UIApplication mdc_safeSharedApplication].statusBarFrame.size.height;
```
Comment (romoore) [inline components/FlexibleHeader/src/private/MDCStatusBarShifter.m:87]: Nit: Consider using [`UIApplicationStatusBarFrameUserInfoKey`](https://developer.apple.com/documentation/uikit/uiapplicationdidchangestatusbarframenotification?language=objc) from the notification.

File: components/FlexibleHeader/src/private/MDCStatusBarShifter.m [hunk 10]
```diff
@@ -205,12 +223,12 @@ - (void)setOffset:(CGFloat)offset {
     return;
   }
 
-  // Bound the status bar range to [0...kStatusBarExpectedHeight].
-  CGFloat statusOffsetY = MIN(kStatusBarExpectedHeight, offset);
+  // Bound the status bar range to [0...statusBarHeight].
```
Comment (romoore) [inline components/FlexibleHeader/src/private/MDCStatusBarShifter.m]: Question: Should this comment match the variable (`_originalStatusBarHeight`) or is it intentionally different?

File: components/FlexibleHeader/src/private/MDCStatusBarShifter.m [hunk 11]
```diff
@@ -242,9 +260,7 @@ - (void)setEnabled:(BOOL)enabled {
 }
 
 - (BOOL)canUpdateStatusBarFrame {
-  CGRect statusBarFrame = [[UIApplication mdc_safeSharedApplication] statusBarFrame];
-  CGFloat statusBarHeight = MIN(statusBarFrame.size.width, statusBarFrame.size.height);
-  return ((statusBarHeight == kStatusBarExpectedHeight) || _statusBarReplicaView ||
+  return (![[UIApplication mdc_safeSharedApplication] isStatusBarHidden] || _statusBarReplicaView ||
```
Comment (romoore) [inline components/FlexibleHeader/src/private/MDCStatusBarShifter.m]: `isStatusBarHidden` [is deprecated](https://developer.apple.com/documentation/uikit/uiapplication/1622982-isstatusbarhidden), do we want to use it anyway?

**Review by romoore — APPROVED** [review]
Looks like it adjusts the scroll insets correctly. Do we plan a second PR to fix the left/right/bottom safe area insets?

**Comment by chuga** [comment]
Why would the flexible header need to "fix the left/right/bottom safe area insets"? Fix what exactly?

**Comment by romoore** [comment]
Maybe I misunderstood, but I thought that the FHV and FHVC should account for the safe area for any subviews or child VCs? And make sure that the bottom doesn't extend beyond the safe area?

**Comment by chuga** [comment]
Since FHV and FHVC are not the owners of the scroll view (it isn't a subview, we just track it), we shouldn't be trying to do anything with it itself beyond playing nicely with any safe area insets it might come with.

**Comment by ianegordon** [comment]
Hi @chuga can we verify this works as expected on the new Xcode 9,1b?

File: components/FlexibleHeader/src/private/MDCStatusBarShifter.m [hunk 12]
```diff
@@ -56,21 +55,38 @@ @implementation MDCStatusBarShifter {
 
   BOOL _prefersStatusBarHidden;
   MDCStatusBarShifterState _snapshotState;
+
+  CGFloat _originalStatusBarHeight;
```
Comment (brianjmoore) [inline components/FlexibleHeader/src/private/MDCStatusBarShifter.m]: Documentation comment?

**Review by brianjmoore — COMMENTED** [review]
Has this been tested with an iOS 11 deployment target/minimum OS? That might help flush out more deprecated API usage and it would be nice to build cleanly for apps that have decided to leave iOS 10 and earlier behind (which I suspect will be a popular option)

Overall the approach looks reasonable to me. Can't wait to try it out.

File: components/FlexibleHeader/examples/FlexibleHeaderHorizontalPagingExample.m (hunk 6 above)
Comment (chuga) [reply components/FlexibleHeader/examples/FlexibleHeaderHorizontalPagingExample.m:149]: Yup, it was. The headerView expects this to be set to YES if there's more one scroll view that it's supposed to be tracking.

File: components/FlexibleHeader/examples/supplemental/FlexibleHeaderTopLayoutGuideSupplemental.m (hunk 7 above)
Comment (chuga) [reply components/FlexibleHeader/examples/supplemental/FlexibleHeaderTopLayoutGuideSupplemental.m]: Done

File: components/FlexibleHeader/src/private/MDCStatusBarShifter.m (hunk 12 above)
Comment (chuga) [reply components/FlexibleHeader/src/private/MDCStatusBarShifter.m]: Done

File: components/FlexibleHeader/src/private/MDCStatusBarShifter.m (hunk 9 above)
Comment (chuga) [reply components/FlexibleHeader/src/private/MDCStatusBarShifter.m:87]: I gave that a try, but it doesn't work as expected on an iPhone X :/
This is what I get when I switch from portrait to landscape (the correct value is 0):
screen shot 2017-09-28 at 4 03 31 pm

File: components/FlexibleHeader/src/private/MDCStatusBarShifter.m (hunk 10 above)
Comment (chuga) [reply components/FlexibleHeader/src/private/MDCStatusBarShifter.m]: Fixed

File: components/FlexibleHeader/src/private/MDCStatusBarShifter.m (hunk 11 above)
Comment (chuga) [reply components/FlexibleHeader/src/private/MDCStatusBarShifter.m]: Bummer, that getter is really useful. Fixed.

**Comment by chuga** [comment]
I've tested it on Xcode 8 and also on 9 with an iOS 11 deployment target/minimum OS, things seem to be working as expected. I can't download iOS 8 for some weird reason (see screenshot), so if you have it help me out and take this for a spin.

screen shot 2017-09-28 at 4 32 12 pm

I'll do a final check on Xcode 9.1 / iOS 11.1 before I merge this in. If you have any other comments, now is the time to share them.

Thanks for reviewing!

File: components/FlexibleHeader/src/MDCFlexibleHeaderView.m [hunk 13]
```diff
@@ -468,6 +517,42 @@ - (MDCFlexibleHeaderScrollViewInfo *)fhv_addInsetsToScrollView:(UIScrollView *)s
   return info;
 }
 
+- (void)fvh_adjustInsetsForSafeAreaInScrollView:(UIScrollView *)scrollView {
```
Comment (jverkoey) [inline components/FlexibleHeader/src/MDCFlexibleHeaderView.m]: super minor: `fhv`

**Review by jverkoey — APPROVED** [review]

**Comment by romoore** [comment]
Tested Xcode 8/iOS 8.3/iPad Air, all the examples under "Flexible Header" seem fine.

**Comment by chuga** [comment]
Thank you Rob
